### DISCUSSION TITLE
Security hardening of distributed cache

### DISCUSSION THREAD

**Discussion by Doctor-love** [keycloak/keycloak#16847] [Ideas]
Greetings, contributors!

After setting up and managing a Keycloak cluster for a while, I find that the scariest part (since migrating away from WildFly, anyway) is restring access and securing Infinispan communication.  
  
As I've understood it, the default configuration does not enforce any authentication nor transport level encryption.
According to the documentation, production Keycloak environments should be placed on an isolated network to mitigate these and other issues. While relying on network traffic restrictions for securing systems is [generally considered a bad practice these days](https://en.wikipedia.org/wiki/Zero_trust_security_model), this should be especially true for a critical high-trust service such as an IdP.

I've spent quite a bit of time to harden the Infinispan configuration by creating a custom JGroups stack (inspired by [the upstream documentation](https://infinispan.org/docs/stable/titles/security/security.html#configuring-jgroups-asym-encrypt_secure-cluster-transport)), which required lots of it trial and error unfortunately. I think there would be lots of value if the project developers, as domain experts, provided guidance related to this.

As I see it, there are a few things that could be improved:
- Clearly document the security implications of the default Infinipan configuration
- Provide one or two examples of hardened cache.xml example files (with token, TLS key exchange and/or similar)
- Change the defaults to be usable without relying on network traffic restrictions for security

Suggested topics for the discussion thread:
- What do the community think about this - do you share my concern?
- What would be a good baseline and what could suite the utmost paranoid?
- Care to share your/example hardened cache configuration?

Keep up the good work and thanks for your effort!

**Comment by jerrac** [comment]
@Doctor-love So, I've been in dev mode with Keycloak for a while now and now I want to push out to test. Part of that is getting the cluster working, which as far as I can tell, works via the infinispan cache. Which lead me to trying to figure out how infinispan knows that a new node is authorized, and eventually to this post.

So, I take it that if someone could get a keycloak instance running on the same network as yours, then infinispan would happily let it join the cluster?

How is your jgroups method of securing things working? Any tips you could share?

**Reply by Doctor-love** [reply]
Did you check out https://infinispan.org/docs/stable/titles/security/security.html#configuring-jgroups-asym-encrypt_secure-cluster-transport ?

**Reply by jerrac** [reply]
Ok, so, if I'm reading that correctly, you can configure a keystore that Infinispan will use to look up a CA cert to verify that the other instance of infinispan is using a cert signed by that CA cert? 

I feel like that isn't quite right. But maybe close?

So, I'd generate a CA cert, add it to the keystore, then for each Keycloak instance, generate a cert from that CA and give each instance a version of the keystore with the cert and key in in. Would infinispan need the certs to have the correct hostnames? Or can I just have a signed cert?

'Cause I'll be using Docker, and it's never fun getting ssl certs configured to work properly when dealing with replicated services if they need specific hostnames....

**Reply by Doctor-love** [reply]
Yep - that's basically it AFAIK. Any certificate signed by the specified CA will be trusted and "hostnames" do not need to match. You don't need certificates from a publicly trusted CA, so any tool like cfssl or easy-rsa should do it.

Another option if you don't wanna deal with certificates is to use symmetric encryption instead - it's described further up on the same page I referenced.

**Reply by jerrac** [reply]
Ok, so I've been trying to get the symmetric version working. I can see that my isbn config is getting read in the debug logs, but it is not outputting the `[org.infinispan.CLUSTER] ISPN000078: Starting JGroups channel cluster with stack <encrypted_stack_name>` message the infinispan docs say to expect. I get the default "[org.infinispan.CLUSTER] (keycloak-cache-init) ISPN000078: Starting JGroups channel `ISPN`" message.

What I'm wondering is if I'm putting the keystore file in the correct location. I haven't been able to find anything on where the embedded version of inifinispan would set it's `$ISPN_HOME` to. So far I've tried the root keycloak dir, and the keycloak/conf dir with no luck.

Would anyone want to chime in with where that directory is?

**Reply by jerrac** [reply]
I had been updating the config when the container launched, assuming Keycloak/Infinispan would read the config files at launch for any new values. Since they don't do that, they were reading the "REPLACE_*" values I was using in the Docker image, which is why Infinispan kept defaulting the jgroup. Once I figured that out, I moved the `kc.sh build` command into my startup script after I replaced the placeholder values. That got me to the point of Infinispan trying to use the configured jgroup.

But, I'm still screwing up somewhere. Specifically Keycloak is  not able to decrypt any messages. I get errors like:

```
keycloak-temp  | 2024-02-22 16:47:35,155 ERROR [org.jgroups.protocols.SYM_ENCRYPT] (jgroups-6,keycloak-temp.blah.tld) keycloak-temp.blah.tld: failed decrypting message from keycloak.blah.tld (offset=0, length=64, buf.length=64): java.security.InvalidKeyException: No installed provider supports this key: com.sun.crypto.provider.PBEKey, headers are GMS: GmsHeader[JOIN_RSP], UNICAST3: DATA, seqno=1, first, SYM_ENCRYPT: EncryptHeader [version=DDA28D1536908AEB1DE2DF6A938F6ACA], TP: [cluster=keycloak-cache-cluster]
```

I generated my keystore like this:
```
keytool -genseckey -keystore devcloak.p12 -alias infinispan_key -keyalg PBEWithHmacSHA256AndAES_256 -validity 3650 -storetype PKCS12
```

Both my main Keycloak instance and my -temp instance are using the same keystore. As in it's bind mounted, and the startup script copies it to where infinispan is configured to look for it.

I'm guessing that I'm not using the correct algorithm, or something, when generating the keystore. But I couldn't find any information from Keycloak or Infinispan on what algorithm to use, or what other settings I might be getting wrong.

Anyone have any tips?

Thanks!

FYI, here is my jgroups config:
```xml
<jgroups>
    <!-- Creates a secure JGroups stack named "encrypt-tcp" that extends the default TCP stack. -->
    <stack name="encrypt-tcp" extends="tcp">
      <!-- Adds a keystore from which nodes obtain secret keys. -->
      <!-- Uses the stack.combine and stack.position attributes to insert SYM_ENCRYPT into the default TCP stack after VERIFY_SUSPECT2. -->
      <SYM_ENCRYPT keystore_name="/REPLACE_INFINI_KEYSTORE_NAME.p12"
                   keystore_type="PKCS12"
                   sym_algorithm="PBEWithHmacSHA256AndAES_256"
                   store_password="REPLACE_INFINI_KEYSTORE_PASS"
                   key_password="REPLACE_INFINI_KEYSTORE_PASS"
                   alias="REPLACE_INFINI_KEYSTORE_KEY_ALIAS"
                   stack.combine="INSERT_AFTER"
                   stack.position="VERIFY_SUSPECT2"/>
    </stack>
  </jgroups>
```

Edit:
I found the clue I needed in the 'lib/lib/main/org.jgroups.jgroups-*.Final.jar` file. The `sym-encrypt.xml` file listed the possible algorithms, and a couple failed keytool commands based on that resulted in a successful command and a keystore file that worked. :)

**Reply by vbrik** [reply]
Just in case anybody else is also struggling with this. This is the command that seems to have worked for me: `keytool -genseckey -alias ispn -storetype PKCS12 -keyalg AES -keysize 256 -keystore ispn.p12 -validity 36500`

When using `PBEWithHmacSHA256AndAES_256` I got an exception about PBE parameters that I couldn't figure out how to fix.
